File: ocaml/day7.ml
open Base

let max_int = List.reduce_exn ~f:Int.max

type input = int list

(* Parse input into a list of numbers *)
let parse_input lines =
  List.hd_exn lines |> String.split ~on:',' |> List.map ~f:Int.of_string

(* Build a table with the cost for travelling distance i at each index *)
let make_fuel_table max =
  let arr = Array.create ~len:(max + 1) 0 in
  for i = 1 to max do
    arr.(i) <- arr.(i - 1) + i
  done;
  arr

(* Calculate the fuel to move all crabs to a given position *)
let calc_fuel crabs cost_fn pos =
  List.map crabs ~f:(fun p -> Int.abs (pos - p) |> cost_fn)
  |> List.fold ~init:0 ~f:( + )

(* Calculate minimum fuel using a cost function f *)
let calc_min_fuel crabs f =
  List.range 0 (max_int crabs + 1)
  |> List.map ~f:(calc_fuel crabs f)
  |> List.reduce_exn ~f:Int.min

(* Part 1: Cost is just the identity function *)
let part1 crabs = calc_min_fuel crabs Fn.id

(* Part 2: Build a lookup table with fuel costs *)
let part2 crabs =
  let fuel_table = make_fuel_table (max_int crabs) in
  calc_min_fuel crabs (fun n -> fuel_table.(n))

File: haskell/day7.hs
module Day7 where

import Util
import qualified Aoc
import Data.Array

-- Parse input into a list of numbers
parse :: [String] -> [Int]
parse lines = map read . splitList ',' $ head lines

-- Build a table with the cost of travelling distance i at each index
makeFuelTable :: Int -> Array Int Int
makeFuelTable n = array (0, n) (zip [0..] costs)
  where costs = scanl (+) 0 [1..n]

-- Calculate the fuel to move all crabs to a given position
calcFuel :: [Int] -> (Int -> Int) -> Int -> Int
calcFuel crabs f pos = sum $ map (f . abs . (pos `subtract`)) crabs

-- Calculate minimum fuel using a cost function f
calcMinFuel :: [Int] -> (Int -> Int) -> Int
calcMinFuel crabs f = minimum [calcFuel crabs f pos | pos <- [0..maxPos]]
  where maxPos = maximum crabs + 1

-- Part 1: Cost is just the identity function
part1 crabs = calcMinFuel crabs id

-- Part 2: Build a lookup table with fuel costs
part2 crabs = calcMinFuel crabs (t !)
  where t = makeFuelTable (maximum crabs + 1)

run = Aoc.run 7 parse part1 part2

